README
This is libpbo, version 0.2.1
A library to work with Arma PBO files

Quickstart:
run ./autogen.sh to setup the autohell
afterwards, usual ./configure ; make ; make install stuff applies

See INSTALL for generic autohell compile/install instructions.

(C) 2015 Emir Marincic <>
Contact: 

See COPYING for license information


NEWS
2015/11/15 - 0.0.0 started
2015/11/22 - 0.1.0 done
2015/11/24 - 0.2.0 done
2015/11/24 - 0.2.1 done

libpbo/pbo.c
/* pbo.c - (C) 2015, Emir Marincic
 * libpbo - A library to work with Arma PBO files
 * See README for contact-, COPYING for license information. */

#if HAVE_CONFIG_H
# include <config.h>
#endif

#include <stdio.h>
#include <stdint.h>
#include <stddef.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include "sha.h"

#include <libpbo/pbo.h>

#define MAXNAMELEN 512

#define WRITE_N_SHA(P,S,N,F,C) \
    fwrite((P), (S), (N), (F)); \
    SHA1Input((C), (uint8_t *)(P), (S) * (N));

typedef enum
{
    CLEAR = 0,
    EXISTING,
    NEW,
} pbo_state;

enum{
    PACKING_METHOD = 0,
    ORIGINAL_SIZE,
    RES,
    TIME_STAMP,
    DATA_SIZE,
};

struct header_extension {
    size_t len;
    char **entries;
};

struct pbo_entry {
    char *name;
    uint32_t properties[5];
    struct header_extension *ext;
    size_t file_offset;
    unsigned char *data;
};

struct list_entry {
    struct list_entry *next;
    struct pbo_entry *data;
};

struct pbo {
    size_t headersz;
    struct list_entry *root;
    struct list_entry *last;
    char *filename;
    pbo_state state;
};

static pbo_error pbo_add_header_extension(struct header_extension *he, const char *e);
static pbo_error pbo_list_add_entry(pbo_t d, struct pbo_entry *pe);
static void pbo_clear_list(pbo_t d);
static struct list_entry *pbo_find_file(pbo_t d, const char *file);
static int pbo_util_getdelim(unsigned char *dst, FILE *src, size_t dstsz, char delim);
static char *pbo_util_strdup(const char *src);

pbo_t pbo_init(const char *filename)
{
    struct pbo *d = malloc(sizeof *d);
    if(!d)
        goto cleanup; //Malloc Error

    d->filename = pbo_util_strdup(filename);
    if(!d->filename)
        goto cleanup;

    d->root = NULL;
    d->last = NULL;
    d->headersz = 0;
    d->state = CLEAR;
    return d;

cleanup:
    free(d);
    return NULL;
}

void pbo_clear(pbo_t d)
{
    if(!d)
        return;

    pbo_clear_list(d);

    free(d->filename);
    d->filename = NULL;
    d->headersz = 0;
    d->state = CLEAR;
}

void pbo_dispose(pbo_t d)
{
    if(!d)
        return;
    pbo_clear(d);
    free(d);
}

pbo_error pbo_set_filename(pbo_t d, const char *filename)
{
    if(!d)
        return PBO_ERROR_NEXIST;
    if(d->state == EXISTING)
        return PBO_ERROR_STATE;

    free(d->filename);
    d->filename = NULL;
    d->filename = pbo_util_strdup(filename);

    if(!d->filename)
        return PBO_ERROR_MALLOC;
    return PBO_SUCCESS;
}

pbo_error pbo_read_header(pbo_t d)
{
    if(!d)
        return PBO_ERROR_NEXIST;
    if(d->state != CLEAR)
        return PBO_ERROR_STATE;

    FILE *file = fopen(d->filename, "r");
    if(!file)
        return PBO_ERROR_IO; //I/O Error

    char buf[MAXNAMELEN];
    size_t file_offset = 0;

    struct pbo_entry *pe = NULL;
    for(int i = 0;; i++) {
        int sz = pbo_util_getdelim(buf, file, sizeof buf, '\0');
        if(sz < 0)
            return PBO_ERROR_BROKEN; //Broken Pbo header

        pe = malloc(sizeof *pe);
        if(!pe)
            goto cleanup; //Malloc error

        pe->ext = NULL;
        pe->data = NULL;

        pe->name = pbo_util_strdup(buf);
        if(!pe->name)
            goto cleanup;

        fread(pe->properties, 4, 5, file); //Get all properties
        pe->file_offset = file_offset;
        file_offset += pe->properties[DATA_SIZE];
        if(!sz && !i) { //Header Extension
            pe->ext = malloc(sizeof *pe->ext);
            if(!pe->ext)
                goto cleanup; //Malloc Error

            pe->ext->len = 0;
            pe->ext->entries = NULL;

            //Add Header extension can fail should handle that
            while(pbo_util_getdelim(buf, file, sizeof buf, '\0') > 0)
                pbo_add_header_extension(pe->ext, buf);
            pbo_add_header_extension(pe->ext, "\0");
        }

        if(pbo_list_add_entry(d, pe))
            goto cleanup;

        if(!sz && i)
            break;
    }
    d->headersz = ftell(file);
    d->state = EXISTING;
    fclose(file);
    return PBO_SUCCESS;

cleanup:
    if(pe){
        free(pe->name);
        free(pe->ext);
    }
    free(pe);
    fclose(file);
    return PBO_ERROR_MALLOC;
}

pbo_error pbo_write(pbo_t d)
{
    if(!d)
        return PBO_ERROR_NEXIST;
    if(d->state != NEW)
        return PBO_ERROR_STATE;

    FILE *file = fopen(d->filename, "w");
    if(!file)
        return PBO_ERROR_IO;

    //Add the dummy entry to indicate end of header
    struct pbo_entry *pe = malloc(sizeof *pe);
    if(!pe)
        goto cleanup; //Malloc Error

    pe->name = pbo_util_strdup("");
    if(!pe->name)
        goto cleanup;

    pe->properties[PACKING_METHOD] = 0;
    pe->properties[ORIGINAL_SIZE] = 0;
    pe->properties[RES] = 0;
    pe->properties[TIME_STAMP] = 0;
    pe->properties[DATA_SIZE] = 0;
    pe->file_offset = 0;
    pe->data = NULL;
    pe->ext = NULL;
    if(pbo_list_add_entry(d, pe))
        goto cleanup;

    SHA1Context ctx;
    SHA1Reset(&ctx);

    //First write the header
    for(struct list_entry *e = d->root; e; e = e->next) {
        WRITE_N_SHA(e->data->name, 1, strlen(e->data->name) + 1, file, &ctx);
        WRITE_N_SHA(e->data->properties, 4, 5, file, &ctx);
        if(e->data->ext) {
            for(int i = 0; i < e->data->ext->len; i++) {
                WRITE_N_SHA(e->data->ext->entries[i], 1, strlen(e->data->ext->entries[i]) + 1, file, &ctx);
            }
        }
    }

    //Then write the data block
    for(struct list_entry *e = d->root; e; e = e->next){
        if(*e->data->name != '\0'){
            WRITE_N_SHA(e->data->data, 1, e->data->properties[DATA_SIZE], file, &ctx);
        }
    }

    //Finalize SHA and write it at the end
    uint8_t sha[SHA1HashSize];
    SHA1Result(&ctx, sha);
    fputc('\0', file); //Format specifies a null before the hash
    fwrite(sha, 1, SHA1HashSize, file);

    fclose(file);
    return PBO_SUCCESS;

cleanup:
    fclose(file);
    if(pe)
        free(pe->name);
    free(pe);
    return PBO_ERROR_MALLOC;
}

//TODO: Add some other way of reading files. Possibly a fread like API.
size_t pbo_read_file(pbo_t d, const char *filename, void *buf, size_t size)
{
    if(!d || !filename || d->state != EXISTING)
        return 0;

    struct list_entry *e = pbo_find_file(d, filename);
    if(!e)
        return 0; //Doesn't exist

    if(e->data->properties[DATA_SIZE] > size)
        return 0; //Doesn't fit

    FILE *file = fopen(d->filename, "r");
    if(!file)
        return 0; //I/O Error

    fseek(file, e->data->file_offset + d->headersz, SEEK_SET);
    size_t sz = fread(buf, 1, e->data->properties[DATA_SIZE], file);
    fclose(file);
    return sz;
}

pbo_error pbo_init_new(pbo_t d)
{
    if(!d)
        return PBO_ERROR_NEXIST;
    if(d->state != CLEAR)
        return PBO_ERROR_STATE;

    pbo_clear(d);
    d->state = NEW;
    return PBO_SUCCESS;
}

pbo_error pbo_add_extension(pbo_t d, const char *e)
{
    if(!d)
        return PBO_ERROR_NEXIST;
    if(d->state != NEW)
        return PBO_ERROR_STATE;

    struct pbo_entry *pe = NULL;
    if(*d->root->data->name != '\0') {
        //Add the dummy entry with the extension
        pe = malloc(sizeof *pe);
        if(!pe)
            goto cleanup; //Malloc Error
    
        pe->name = pbo_util_strdup("");
        pe->properties[PACKING_METHOD] = 0x56657273;
        pe->properties[ORIGINAL_SIZE] = 0;
        pe->properties[RES] = 0;
        pe->properties[TIME_STAMP] = 0;
        pe->properties[DATA_SIZE] = 0;
    
        pe->ext = malloc(sizeof *pe->ext);
        if(!pe->ext)
            goto cleanup; //Malloc Error

        pe->data = NULL;

        //Insert the header extension entry at the beginning
        struct list_entry *le = malloc(sizeof *le);
        if(!le)
            goto cleanup; //Malloc Error

        le->next = d->root;
        le->data = pe;
        d->root = le;
    }

    return pbo_add_header_extension(d->root->data->ext, e);

cleanup:
    if(pe)
        free(pe->ext);
    free(pe);
    return PBO_ERROR_MALLOC;
}

pbo_error pbo_add_file_d(pbo_t d, const char *name, void *data,  size_t size)
{
    if(!d)
        return PBO_ERROR_NEXIST;
    if(d->state != NEW)
        return PBO_ERROR_STATE;

    struct pbo_entry *pe = malloc(sizeof *pe);
    if(!pe)
        goto cleanup; //Malloc Error

    pe->name = pbo_util_strdup(name);
    if(!pe->name)
        goto cleanup; //Malloc Error

    pe->data = malloc(size);
    if(!pe->data)
        goto cleanup; //Malloc Error

    memcpy(pe->data, data, size);

    pe->properties[PACKING_METHOD] = 0;
    pe->properties[ORIGINAL_SIZE] = size;
    pe->properties[RES] = 0;
    pe->properties[TIME_STAMP] = (uint32_t)time(NULL);
    pe->properties[DATA_SIZE] = size;

    pe->ext = NULL;

    return pbo_list_add_entry(d, pe);

cleanup:
    if(pe)
        free(pe->name);
    free(pe);
    return PBO_ERROR_MALLOC;
}

pbo_error pbo_add_file_f(pbo_t d, const char *name, FILE *file)
{
    if(!d || !file)
        return PBO_ERROR_NEXIST;
    if(d->state != NEW)
        return PBO_ERROR_STATE;

    struct pbo_entry *pe = malloc(sizeof *pe);
    if(!pe)
        goto cleanup;

    pe->name = pbo_util_strdup(name);
    if(!pe->name)
        goto cleanup;

    //Get file size
    fseek(file, 0, SEEK_END);
    size_t filesz = ftell(file);
    rewind(file);

    pe->data = malloc(filesz);
    if(!pe->data)
        goto cleanup;

    fread(pe->data, 1, filesz, file);
    fclose(file);

    pe->properties[PACKING_METHOD] = 0;
    pe->properties[ORIGINAL_SIZE] = filesz;
    pe->properties[RES] = 0;
    pe->properties[TIME_STAMP] = (uint32_t)time(NULL);
    pe->properties[DATA_SIZE] = filesz;

    pe->ext = NULL;

    return  pbo_list_add_entry(d, pe);

cleanup:
    if(pe)
        free(pe->name);
    free(pe);
    return PBO_ERROR_MALLOC;
}

pbo_error pbo_add_file_p(pbo_t d, const char *name, const char *path)
{
    FILE *file = fopen(path, "r");
    if(!file)
        return PBO_ERROR_IO;
    return pbo_add_file_f(d, name, file);
}

void pbo_get_file_list(pbo_t d, pbo_listcb cb, void *user)
{
    if(!d)
        return;

    for(struct list_entry *e = d->root; e; e = e->next)
        cb(e->data->name, user);
}

size_t pbo_get_file_size(pbo_t d, const char *filename)
{
    if(!d)
        return 0;

    struct list_entry *e = pbo_find_file(d, filename);
    if(!e)
        return 0; //Doesn't Exist

    return e->data->properties[DATA_SIZE];
}

pbo_error pbo_write_to_file(pbo_t d, const char *filename, FILE *file)
{
    if(!d)
        return PBO_ERROR_NEXIST;

    struct list_entry *le = pbo_find_file(d, filename);
    if(!le)
        return PBO_ERROR_NEXIST; //Doesn't exist

    FILE *f = fopen(d->filename, "r");
    if(!f)
        return PBO_ERROR_IO;

    fseek(f, le->data->file_offset + d->headersz, SEEK_SET);
    size_t sz = le->data->properties[DATA_SIZE];
    char buf[sz];
    fread(buf, 1, sz, f);
    fwrite(buf, 1, sz, file);
    fclose(f);
}

void pbo_dump_header(pbo_t d)
{
    if(!d)
        return;

    int i = 0;
    for(struct list_entry *e = d->root; e; e = e->next) {
        struct pbo_entry *pe = e->data;
        printf("Entry(%d): %s\n", i, pe->name);
        for(int i = 0; i <= DATA_SIZE; i++)
            printf("\tproperties[%d] = %d\n", i, pe->properties[i]);
        if(pe->ext) {
            printf("\tHeaderExtension:\n");
            for(int i = 0; i < pe->ext->len; i++)
                printf("\t\tHEntry: %s\n", pe->ext->entries[i]);
        }
        i++;
    }
}

static pbo_error pbo_add_header_extension(struct header_extension *he, const char *e)
{
    if(he->len % 4 == 0) {
        char **new = realloc(he->entries, (he->len + 4) * sizeof *new);
        if(!new)
            return PBO_ERROR_MALLOC; //Malloc Error
        he->entries = new;
    }
    he->entries[he->len] = pbo_util_strdup(e);
    if(!he->entries[he->len])
        goto cleanup;
    he->len++;

    return PBO_SUCCESS;

cleanup:
    if(he->len % 4 == 0)
        he->entries = realloc(he->entries, (he->len) * sizeof *new);
    return PBO_ERROR_MALLOC;
}

static pbo_error pbo_list_add_entry(pbo_t d, struct pbo_entry *pe)
{
    if(!d)
        return PBO_ERROR_NEXIST;

    struct list_entry *le = malloc(sizeof *le);
    if(!le)
        return PBO_ERROR_MALLOC; //Malloc Error

    le->next = NULL;
    le->data = pe;

    if(!d->root)
        d->root = le;
    else
        d->last->next = le;
    d->last = le;

    return PBO_SUCCESS;
}

static void pbo_clear_list(pbo_t d)
{
    struct list_entry *e = d->root;
    while(e) {
        struct list_entry *t = e->next;
        free(e->data->name);
        if(e->data->ext) {
            for(int i = 0; i < e->data->ext->len; i++)
                free(e->data->ext->entries[i]);
            free(e->data->ext->entries);
            free(e->data->ext);
        }
        free(e->data->data);
        free(e->data);
        free(e);
        e = t;
    }
    d->root = NULL;
    d->last = NULL;
}

static struct list_entry *pbo_find_file(pbo_t d, const char *file)
{
    if(!d)
        return NULL;

    for(struct list_entry *e = d->root; e; e = e->next)
        if(!strcmp(e->data->name, file))
            return e;
    return NULL;
}

static int pbo_util_getdelim(unsigned char *dst, FILE *src, size_t dstsz, char delim)
{
    if(!dstsz || !dst)
        return -1;

    char c = 0;
    int sz = 0;

    while(dstsz && (c = fgetc(src)) != EOF && c != delim)
        sz++, dstsz--, *dst++ = c;

    if(dstsz && !feof(src))
        *dst = delim;
    else
        sz = -1;
    return sz;
}

static char *pbo_util_strdup(const char *src)
{
    size_t len = strlen(src) + 1;
    char *new = malloc(len);

    if(!new)
        return NULL;
    return (char *) memcpy(new, src, len);
}

src/testlibpbo.c
/* testlibpbo.c - (C) 2015, Emir Marincic
 * libpbo - A library to work with Arma PBO files
 * See README for contact-, COPYING for license information. */
 
#if HAVE_CONFIG_H
# include <config.h>
#endif

#include <stddef.h>
#include <stdio.h>
#include <string.h>
#include <sys/stat.h>

#ifdef HAVE_DIRECT_H
    #include <direct.h>
#endif

#include <libpbo/pbo.h>

void create_directories(char *path)
{    
    for(int i = 0; path[i] != '\0'; i++) {
        if(path[i] == '/') {
            path[i] = '\0';
            #ifdef HAVE_DIRECT_H
                mkdir(path);
            #else
                mkdir(path, 0755);
            #endif
            path[i] = '/';
        }
    }
}

void list_callback(const char *filename, void *user)
{
    if(filename == '\0')
        return;

    pbo_t d = (pbo_t)user;

    char buf[512];
    strcpy(buf, filename);

    for(int i = 0; buf[i] != '\0'; i++)
        if(buf[i] == '\\')
            buf[i] = '/';

    create_directories(buf);

    FILE *file = fopen(buf, "w");
    if(!file)
        return;
    pbo_write_to_file(d, filename, file);
    fclose(file);
}

int main(void)
{
    pbo_t d = pbo_init("read.pbo");
    pbo_read_header(d);
    pbo_dump_header(d);
    pbo_get_file_list(d, list_callback, d);
    pbo_clear(d);
    pbo_init_new(d);
    pbo_set_filename(d, "write.pbo");
    pbo_add_file_p(d, "test.txt", "test.txt");
    pbo_write(d);
    pbo_dispose(d);
    return 0;
}

include/libpbo/pbo.h
/* pbo.h - (C) 2015, Emir Marincic
 * libpbo - A library to work with Arma PBO files
 * See README for contact-, COPYING for license information. */

#ifndef LIBpbo_pbo_H
#define LIBpbo_pbo_H 1

typedef enum
{
    PBO_SUCCESS = 0,
    PBO_ERROR_NEXIST,
    PBO_ERROR_BROKEN,
    PBO_ERROR_MALLOC,
    PBO_ERROR_IO,
    PBO_ERROR_STATE,
} pbo_error;

typedef void (*pbo_listcb)(const char*, void*);

typedef struct pbo *pbo_t;

pbo_t pbo_init(const char *filename);
void pbo_clear(pbo_t d);
void pbo_dispose(pbo_t d);
pbo_error pbo_set_filename(pbo_t d, const char *filename);

pbo_error pbo_read_header(pbo_t d);
pbo_error pbo_write(pbo_t d);

size_t pbo_read_file(pbo_t d, const char *filename, void *buf, size_t size);

pbo_error pbo_init_new(pbo_t d);
pbo_error pbo_add_extension(pbo_t d, const char *e);
pbo_error pbo_add_file_d(pbo_t d, const char *name, void *data,  size_t size);
pbo_error pbo_add_file_f(pbo_t d, const char *name, FILE *file);
pbo_error pbo_add_file_p(pbo_t d, const char *name, const char *path);

void pbo_get_file_list(pbo_t d, pbo_listcb cb, void *user);
size_t pbo_get_file_size(pbo_t d, const char *filename);

pbo_error pbo_write_to_file(pbo_t d, const char *filename, FILE *file);
void pbo_dump_header(pbo_t d);

#endif /* LIBpbo_pbo_H */


libpbo/sha-private.h
/*************************** sha-private.h ***************************/
/********************** See RFC 4634 for details *********************/
#ifndef _SHA_PRIVATE__H
#define _SHA_PRIVATE__H
/*
 * These definitions are defined in FIPS-180-2, section 4.1.
 * Ch() and Maj() are defined identically in sections 4.1.1,
 * 4.1.2 and 4.1.3.
 *
 * The definitions used in FIPS-180-2 are as follows:
 */

#ifndef USE_MODIFIED_MACROS
#define SHA_Ch(x,y,z)        (((x) & (y)) ^ ((~(x)) & (z)))
#define SHA_Maj(x,y,z)       (((x) & (y)) ^ ((x) & (z)) ^ ((y) & (z)))

#else /* USE_MODIFIED_MACROS */
/*
 * The following definitions are equivalent and potentially faster.
 */

#define SHA_Ch(x, y, z)      (((x) & ((y) ^ (z))) ^ (z))
#define SHA_Maj(x, y, z)     (((x) & ((y) | (z))) | ((y) & (z)))
#endif /* USE_MODIFIED_MACROS */

#define SHA_Parity(x, y, z)  ((x) ^ (y) ^ (z))

#endif /* _SHA_PRIVATE__H */

libpbo/sha1.c
/**************************** sha1.c ****************************/
/******************** See RFC 4634 for details ******************/
/*
 *  Description:
 *      This file implements the Secure Hash Signature Standard
 *      algorithms as defined in the National Institute of Standards
 *      and Technology Federal Information Processing Standards
 *      Publication (FIPS PUB) 180-1 published on April 17, 1995, 180-2
 *      published on August 1, 2002, and the FIPS PUB 180-2 Change
 *      Notice published on February 28, 2004.
 *
 *      A combined document showing all algorithms is available at
 *              http://csrc.nist.gov/publications/fips/
 *              fips180-2/fips180-2withchangenotice.pdf
 *
 *      The SHA-1 algorithm produces a 160-bit message digest for a
 *      given data stream.  It should take about 2**n steps to find a
 *      message with the same digest as a given message and
 *      2**(n/2) to find any two messages with the same digest,
 *      when n is the digest size in bits.  Therefore, this
 *      algorithm can serve as a means of providing a
 *      "fingerprint" for a message.
 *
 *  Portability Issues:
 *      SHA-1 is defined in terms of 32-bit "words".  This code
 *      uses <stdint.h> (included via "sha.h") to define 32 and 8
 *      bit unsigned integer types.  If your C compiler does not
 *      support 32 bit unsigned integers, this code is not
 *      appropriate.
 *
 *  Caveats:
 *      SHA-1 is designed to work with messages less than 2^64 bits
 *      long. This implementation uses SHA1Input() to hash the bits
 *      that are a multiple of the size of an 8-bit character, and then
 *      uses SHA1FinalBits() to hash the final few bits of the input.
 */

#include "sha.h"
#include "sha-private.h"

/*
 *  Define the SHA1 circular left shift macro
 */
#define SHA1_ROTL(bits,word) \
                                (((word) << (bits)) | ((word) >> (32-(bits))))

/*
 * add "length" to the length
 */
static uint32_t addTemp;
#define SHA1AddLength(context, length)                     \
        (addTemp = (context)->Length_Low,                      \
         (context)->Corrupted =                                \
                (((context)->Length_Low += (length)) < addTemp) && \
                (++(context)->Length_High == 0) ? 1 : 0)

/* Local Function Prototypes */
static void SHA1Finalize(SHA1Context *context, uint8_t Pad_Byte);
static void SHA1PadMessage(SHA1Context *, uint8_t Pad_Byte);
static void SHA1ProcessMessageBlock(SHA1Context *);

/*
 *  SHA1Reset
 *
 *  Description:
 *      This function will initialize the SHA1Context in preparation
 *      for computing a new SHA1 message digest.
 *
 *  Parameters:
 *      context: [in/out]
 *          The context to reset.
 *
 *  Returns:
 *      sha Error Code.
 *
 */
int SHA1Reset(SHA1Context *context)
{
        if (!context)
                return shaNull;

        context->Length_Low             = 0;
        context->Length_High            = 0;
        context->Message_Block_Index    = 0;

        /* Initial Hash Values: FIPS-180-2 section 5.3.1 */
        context->Intermediate_Hash[0]   = 0x67452301;
        context->Intermediate_Hash[1]   = 0xEFCDAB89;
        context->Intermediate_Hash[2]   = 0x98BADCFE;
        context->Intermediate_Hash[3]   = 0x10325476;
        context->Intermediate_Hash[4]   = 0xC3D2E1F0;

        context->Computed   = 0;
        context->Corrupted  = 0;

        return shaSuccess;
}

/*
 *  SHA1Input
 *
 *  Description:
 *      This function accepts an array of octets as the next portion
 *      of the message.
 *
 *  Parameters:
 *      context: [in/out]
 *          The SHA context to update
 *      message_array: [in]
 *          An array of characters representing the next portion of
 *          the message.
 *      length: [in]
 *          The length of the message in message_array
 *
 *  Returns:
 *      sha Error Code.
 *
 */
int SHA1Input(SHA1Context *context,
        const uint8_t *message_array, unsigned length)
{
    if (!length)
        return shaSuccess;

    if (!context || !message_array)
        return shaNull;

    if (context->Computed) {
        context->Corrupted = shaStateError;
        return shaStateError;
    }

    if (context->Corrupted)
         return context->Corrupted;

    while (length-- && !context->Corrupted) {
        context->Message_Block[context->Message_Block_Index++] =
            (*message_array & 0xFF);

        if (!SHA1AddLength(context, 8) &&
            (context->Message_Block_Index == SHA1_Message_Block_Size))
            SHA1ProcessMessageBlock(context);

        message_array++;
    }

    return shaSuccess;
}

/*
 * SHA1FinalBits
 *
 * Description:
 *   This function will add in any final bits of the message.
 *
 * Parameters:
 *   context: [in/out]
 *     The SHA context to update
 *   message_bits: [in]
 *     The final bits of the message, in the upper portion of the
 *     byte. (Use 0b###00000 instead of 0b00000### to input the
 *     three bits ###.)
 *   length: [in]
 *     The number of bits in message_bits, between 1 and 7.
 *
 * Returns:
 *   sha Error Code.
 */
int SHA1FinalBits(SHA1Context *context, const uint8_t message_bits,
        unsigned int length)
{
    uint8_t masks[8] = {
            /* 0 0b00000000 */ 0x00, /* 1 0b10000000 */ 0x80,
            /* 2 0b11000000 */ 0xC0, /* 3 0b11100000 */ 0xE0,
            /* 4 0b11110000 */ 0xF0, /* 5 0b11111000 */ 0xF8,
            /* 6 0b11111100 */ 0xFC, /* 7 0b11111110 */ 0xFE
    };
    uint8_t markbit[8] = {
            /* 0 0b10000000 */ 0x80, /* 1 0b01000000 */ 0x40,
            /* 2 0b00100000 */ 0x20, /* 3 0b00010000 */ 0x10,
            /* 4 0b00001000 */ 0x08, /* 5 0b00000100 */ 0x04,
            /* 6 0b00000010 */ 0x02, /* 7 0b00000001 */ 0x01
    };

    if (!length)
        return shaSuccess;

    if (!context)
        return shaNull;

    if (context->Computed || (length >= 8) || (length == 0)) {
        context->Corrupted = shaStateError;
        return shaStateError;
    }

    if (context->Corrupted)
         return context->Corrupted;

    SHA1AddLength(context, length);
    SHA1Finalize(context,
        (uint8_t) ((message_bits & masks[length]) | markbit[length]));

    return shaSuccess;
}

/*
 * SHA1Result
 *
 * Description:
 *   This function will return the 160-bit message digest into the
 *   Message_Digest array provided by the caller.
 *   NOTE: The first octet of hash is stored in the 0th element,
 *      the last octet of hash in the 19th element.
 *
 * Parameters:
 *   context: [in/out]
 *     The context to use to calculate the SHA-1 hash.
 *   Message_Digest: [out]
 *     Where the digest is returned.
 *
 * Returns:
 *   sha Error Code.
 *
 */
int SHA1Result(SHA1Context *context,
        uint8_t Message_Digest[SHA1HashSize])
{
    int i;

    if (!context || !Message_Digest)
        return shaNull;

    if (context->Corrupted)
        return context->Corrupted;

    if (!context->Computed)
        SHA1Finalize(context, 0x80);

    for (i = 0; i < SHA1HashSize; ++i)
        Message_Digest[i] = (uint8_t) (context->Intermediate_Hash[i>>2]
                            >> 8 * ( 3 - ( i & 0x03 ) ));

    return shaSuccess;
}

/*
 * SHA1Finalize
 *
 * Description:
 *   This helper function finishes off the digest calculations.
 *
 * Parameters:
 *   context: [in/out]
 *     The SHA context to update
 *   Pad_Byte: [in]
 *     The last byte to add to the digest before the 0-padding
 *     and length. This will contain the last bits of the message
 *     followed by another single bit. If the message was an
 *     exact multiple of 8-bits long, Pad_Byte will be 0x80.
 *
 * Returns:
 *   sha Error Code.
 *
 */
static void SHA1Finalize(SHA1Context *context, uint8_t Pad_Byte)
{
    int i;
    SHA1PadMessage(context, Pad_Byte);
    /* message may be sensitive, clear it out */
    for (i = 0; i < SHA1_Message_Block_Size; ++i)
        context->Message_Block[i] = 0;
    context->Length_Low = 0;  /* and clear length */
    context->Length_High = 0;
    context->Computed = 1;
}

/*
 * SHA1PadMessage
 *
 * Description:
 *   According to the standard, the message must be padded to an
 *   even 512 bits. The first padding bit must be a '1'. The last
 *   64 bits represent the length of the original message. All bits
 *   in between should be 0. This helper function will pad the
 *   message according to those rules by filling the Message_Block
 *   array accordingly. When it returns, it can be assumed that the
 *   message digest has been computed.
 *
 * Parameters:
 *   context: [in/out]
 *     The context to pad
 *   Pad_Byte: [in]
 *     The last byte to add to the digest before the 0-padding
 *     and length. This will contain the last bits of the message
 *     followed by another single bit. If the message was an
 *     exact multiple of 8-bits long, Pad_Byte will be 0x80.
 *
 * Returns:
 *   Nothing.
 */
static void SHA1PadMessage(SHA1Context *context, uint8_t Pad_Byte)
{
    /*
     * Check to see if the current message block is too small to hold
     * the initial padding bits and length. If so, we will pad the
     * block, process it, and then continue padding into a second
     * block.
     */
    if (context->Message_Block_Index >= (SHA1_Message_Block_Size - 8)) {
        context->Message_Block[context->Message_Block_Index++] = Pad_Byte;
        while (context->Message_Block_Index < SHA1_Message_Block_Size)
            context->Message_Block[context->Message_Block_Index++] = 0;

        SHA1ProcessMessageBlock(context);
    } else
        context->Message_Block[context->Message_Block_Index++] = Pad_Byte;

    while (context->Message_Block_Index < (SHA1_Message_Block_Size - 8))
        context->Message_Block[context->Message_Block_Index++] = 0;

    /*
     * Store the message length as the last 8 octets
     */
    context->Message_Block[56] = (uint8_t) (context->Length_High >> 24);
    context->Message_Block[57] = (uint8_t) (context->Length_High >> 16);
    context->Message_Block[58] = (uint8_t) (context->Length_High >> 8);
    context->Message_Block[59] = (uint8_t) (context->Length_High);
    context->Message_Block[60] = (uint8_t) (context->Length_Low >> 24);
    context->Message_Block[61] = (uint8_t) (context->Length_Low >> 16);
    context->Message_Block[62] = (uint8_t) (context->Length_Low >> 8);
    context->Message_Block[63] = (uint8_t) (context->Length_Low);

    SHA1ProcessMessageBlock(context);
}

/*
 * SHA1ProcessMessageBlock
 *
 * Description:
 *   This helper function will process the next 512 bits of the
 *   message stored in the Message_Block array.
 *
 * Parameters:
 *   None.
 *
 * Returns:
 *   Nothing.
 *
 * Comments:
 *   Many of the variable names in this code, especially the
 *   single character names, were used because those were the
 *   names used in the publication.
 */
static void SHA1ProcessMessageBlock(SHA1Context *context)
{
    /* Constants defined in FIPS-180-2, section 4.2.1 */
    const uint32_t K[4] = {
            0x5A827999, 0x6ED9EBA1, 0x8F1BBCDC, 0xCA62C1D6
    };
    int        t;               /* Loop counter */
    uint32_t   temp;            /* Temporary word value */
    uint32_t   W[80];           /* Word sequence */
    uint32_t   A, B, C, D, E;   /* Word buffers */

    /*
     * Initialize the first 16 words in the array W
     */
    for (t = 0; t < 16; t++) {
        W[t]  = ((uint32_t)context->Message_Block[t * 4]) << 24;
        W[t] |= ((uint32_t)context->Message_Block[t * 4 + 1]) << 16;
        W[t] |= ((uint32_t)context->Message_Block[t * 4 + 2]) << 8;
        W[t] |= ((uint32_t)context->Message_Block[t * 4 + 3]);
    }
    for (t = 16; t < 80; t++)
        W[t] = SHA1_ROTL(1, W[t-3] ^ W[t-8] ^ W[t-14] ^ W[t-16]);

    A = context->Intermediate_Hash[0];
    B = context->Intermediate_Hash[1];
    C = context->Intermediate_Hash[2];
    D = context->Intermediate_Hash[3];
    E = context->Intermediate_Hash[4];

    for (t = 0; t < 20; t++) {
        temp = SHA1_ROTL(5,A) + SHA_Ch(B, C, D) + E + W[t] + K[0];
        E = D;
        D = C;
        C = SHA1_ROTL(30,B);
        B = A;
        A = temp;
    }

    for (t = 20; t < 40; t++) {
        temp = SHA1_ROTL(5,A) + SHA_Parity(B, C, D) + E + W[t] + K[1];
        E = D;
        D = C;
        C = SHA1_ROTL(30,B);
        B = A;
        A = temp;
    }

    for (t = 40; t < 60; t++) {
        temp = SHA1_ROTL(5,A) + SHA_Maj(B, C, D) + E + W[t] + K[2];
        E = D;
        D = C;
        C = SHA1_ROTL(30,B);
        B = A;
        A = temp;
    }

    for (t = 60; t < 80; t++) {
        temp = SHA1_ROTL(5,A) + SHA_Parity(B, C, D) + E + W[t] + K[3];
        E = D;
        D = C;
        C = SHA1_ROTL(30,B);
        B = A;
        A = temp;
    }

    context->Intermediate_Hash[0] += A;
    context->Intermediate_Hash[1] += B;
    context->Intermediate_Hash[2] += C;
    context->Intermediate_Hash[3] += D;
    context->Intermediate_Hash[4] += E;

    context->Message_Block_Index = 0;
}